_posts/2023-02-01-user-defined-operators.md

```diff
@@ -0,0 +1,173 @@
+is the support for defining operators on [user defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs).
+Let's take a closer look into the details of the the new mechanism.
+
+### Built-in operators
+Solidity provides the following operators for its built-in types:
+
+| Category            | Operators                                                               |
+|---------------------|-------------------------------------------------------------------------|
+| Bitwise             | `&`, `\|`, `^`, `<<`, `>>`, `~`                                         |
+| Arithmetic          | `+`, `-`, `*`, `/`, `%`, `**`                                           |
+| Comparison          | `==`, `!=`, `<`, `<=`, `>`, `>=`                                        |
+| Boolean             | `!`, `&&`, `||`                                                         |
+| Assignment          | `=`, `+=`, `-=`, `>=`, `*=`, `/=`, `%=`, `&=`, `|=`, `^=`, `<<=`, `>>=` |
+| Increment/decrement | `++`, `--`                                                              |
+| Ternary             | `?:`                                                                    |
+| Other               | `delete`, `new`                                                         |
+
+Most of these operators work only with [value types](https://docs.soliditylang.org/en/latest/types.html#value-types).
+
+In particular, with the exception of `?:` and `=`, these operators are not available on user-defined
+value types, making their use much less convenient than the built-in ones.
+One way to make the use of these types more convenient is attaching helper functions with
+[`using for`](https://docs.soliditylang.org/en/latest/contracts.html#using-for).
+
+### Quick Refresher on `using for`
+`using for` can attach a function to a type.
+
+```solidity
+type Int is int;
+using {add} for Int;
+
+function add(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function test(Int a, Int b) pure returns (Int) {
+    return a.add(b);
+}
+```
+
+Currently it has 3 forms:
+1. `using L for T`: attaches all non-private functions from library `L` to type `T`.
+1. `using L for *`: attaches all non-private functions from library `L` to any type, including types that
+    cannot be named in the language (e.g. types of literals or array slices).
+1. `using {f, g, h} for T`: Attaches functions `f`, `g` and `h` to type `T`.
+
+`using for` can be used at file level or inside contracts, including libraries (but not interfaces).
+
+At file level the directive can be marked as `global` which extends its effect to all source units
+where a type is visible.
+The use of `global` is possible only with UDVTs, structs and enums that are defined in the current source unit.
+
+The following example shows off the main features of `using for` in combination wth UDVTs:
+
```

Comment:
```suggestion
### Quick Refresher on `using for`
The directive `using A for B;` can be used to attach functions (A) as member functions to any type (B). 
These functions will receive the object they are called on as their first parameter.

```solidity
type Int is int;
using {add} for Int;

function add(Int a, Int b) pure returns (Int) {
    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
}

function test(Int a, Int b) pure returns (Int) {
    return a.add(b);
}
```

Currently, it has 3 forms:
1. `using L for T`: attaches all non-private functions from library `L` to type `T`.
1. `using L for *`: attaches all non-private functions from library `L` to any type, including types that
    cannot be named in the language (e.g. types of literals or array slices).
1. `using {f, g, h} for T`: Attaches functions `f`, `g`, and `h` to type `T`.

✅ Can be used:
- at the file level
- inside contracts,  including libraries

❌ Can not be used
- Inside contracts with interfaces

At the file level, the directive can be marked as `global`, which extends its effect to all source units
where a type is visible.
The use of `global` is possible only with UDVTs, structs, and enums that are defined in the current source unit.

The following example shows off the main features of `using for` in combination with UDVTs:

```